homeassistant/components/rotel/media_player.py

```diff
@@ -0,0 +1,493 @@
+    vol.Optional(CONF_SOURCE_ALIASES): vol.Schema(
+        {vol.Any(*ROTEL_RSP1570_SOURCES.keys()): vol.Any(str, None)}),
+})
+
+ATTR_SOURCE_NAME = "source_name"
+ATTR_VOLUME = "volume"
+ATTR_MUTE_ON = "mute_on"
+ATTR_PARTY_MODE_ON = "party_mode_on"
+ATTR_INFO = "info"
+ATTR_ICONS = "icons"
+ATTR_SPEAKER_ICONS = "speaker_icons"
+ATTR_STATE_ICONS = "state_icons"
+ATTR_INPUT_ICONS = "input_icons"
+ATTR_SOUND_MODE_ICONS = "sound_mode_icons"
+ATTR_MISC_ICONS = "misc_icons"
+ATTR_TRIGGERS = "triggers"
+
+ATTR_COMMAND_NAME = "command_name"
+SERVICE_SEND_COMMAND = "rotel_send_command"
+SERVICE_SEND_COMMAND_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.comp_entity_ids,
+    vol.Required(ATTR_COMMAND_NAME): cv.string,
+})
+SERVICE_RECONNECT = "rotel_reconnect"
+SERVICE_RECONNECT_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.comp_entity_ids,
+})
+
+
+async def async_setup_platform(
+        hass,
+        config,
+        async_add_entities,
+        discovery_info=None):
+    """Set up the rsp1570serial platform."""
+    # pylint: disable=unused-argument
+
+    device = RotelMediaPlayer(
+        config.get(CONF_NAME),
+        config.get(CONF_DEVICE),
+        config.get(CONF_SOURCE_ALIASES))
+
+    async def handle_hass_stop_event(event):
+        """Clean up when hass stops."""
+        await device.cleanup()
+
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        handle_hass_stop_event)
+    _LOGGER.debug("Registered device '%s' for HASS stop event", device.name)
+
+    await device.open_connection()
+
+    async_add_entities([device])
+    device.start_read_messages(hass)
+    setup_hass_services(hass)
+
+
+def setup_hass_services(hass):
+    """
+    Register services.
+
+    Note that this function is called for every entity but it
+    only needs to be called once for the platform.
+    It doesn't seem to do any harm but I'd like to tidy that up at some point.
+    """
+    _LOGGER.debug("Setting up hass services")
+
+    async def async_handle_send_command(entity, call):
+        command_name = call.data.get(ATTR_COMMAND_NAME)
+        if isinstance(entity, RotelMediaPlayer):
+            _LOGGER.debug(
+                "%s service sending command %s to entity %s",
+                SERVICE_SEND_COMMAND,
+                command_name,
+                entity.entity_id)
+            await entity.send_command(command_name)
+        else:
+            _LOGGER.debug(
+                "%s service not sending command %s to incompatible entity %s",
+                SERVICE_SEND_COMMAND,
+                command_name,
+                entity.entity_id)
+
+    async def async_handle_reconnect(entity, call):
+        # pylint: disable=unused-argument
+        if isinstance(entity, RotelMediaPlayer):
+            _LOGGER.debug("%s service reconnecting entity %s",
+                          SERVICE_RECONNECT, entity.entity_id)
+            await entity.reconnect()
```

Comment:
Can we make this reconnect automatically without requiring the user to intervene and call the service?